langchain-core/src/runnables/graph_mermaid.ts

```diff
@@ -97,22 +97,44 @@ export function drawMermaid(
 
   const seenSubgraphs = new Set<string>();
 
+  // sort prefixes by path length for correct nesting
+  function sortPrefixesByDepth(prefixes: string[]): string[] {
+    return [...prefixes].sort((a, b) => {
+      return a.split(":").length - b.split(":").length;
+    });
+  }
+
   function addSubgraph(edges: Edge[], prefix: string): void {
     const selfLoop = edges.length === 1 && edges[0].source === edges[0].target;
     if (prefix && !selfLoop) {
       const subgraph = prefix.split(":").pop()!;
-      if (seenSubgraphs.has(subgraph)) {
+
+      if (seenSubgraphs.has(prefix)) {
         throw new Error(
-          `Found duplicate subgraph '${subgraph}' -- this likely means that ` +
+          `Found duplicate subgraph '${subgraph}' at '${prefix} -- this likely means that ` +
             "you're reusing a subgraph node with the same name. " +
             "Please adjust your graph to have subgraph nodes with unique names."
         );
       }
 
-      seenSubgraphs.add(subgraph);
+      seenSubgraphs.add(prefix);
       mermaidGraph += `\tsubgraph ${subgraph}\n`;
     }
 
+    // all nested prefixes for this level, sorted by depth
+    const nestedPrefixes = sortPrefixesByDepth(
```

Comment:
Will need to take a closer look, but shouldn't all prefixes passed into `sortPrefixesByDepth` have the same length?